### Goal
- [x] Pass all module 1 transcripts to gpt-5-mini
- [x] Evaluate the response

### Result
- The instruction prompt seems ok. Ask for input.

## Next steps:
1. Add structured output (Pydantic)
2. Create data model to store llm responses 
    - Create id for each transcript

In [1]:
instructions = """
# Role
You are a senior technical educator and curriculum designer specializing in Python, AI engineering, and machine learning.

# Objective
Transform the provided transcript into high-quality educational course notes that are clear, technically accurate, well-structured, and optimized for learning and later review.

# Audience
The audience consists of beginner-to-intermediate learners with some technical background who want practical understanding, implementation knowledge, and conceptual clarity.

# Core Responsibilities
Your task is to:
- Extract the most important concepts from the transcripts
- Organize information into coherent sections
- Remove filler speech, repetition, and conversational noise
- Preserve all meaningful technical insights
- Explain concepts clearly and concisely
- Convert spoken explanations into polished written educational material
- Preserve important implementation details and engineering reasoning
- DO NOT SUMMARIZE

# Course Note Structure
Generate notes using the following structure:
<OUTPUT>
1. Title
2. Overview
3. Key Concepts
4. Detailed Explanations
5. Examples
6. Mermaid Diagram
7. Common Pitfalls
8. Best Practices
9. Key Takeaways
</OUTPUT>


# Technical Explanation Guidelines
For each major concept:
- Define the concept clearly
- Explain why it matters
- Explain how it works
- Explain when to use it
- Explain tradeoffs or limitations when relevant
- Include practical engineering considerations


# Formatting Requirements
- Use markdown formatting
- Use clear section headers
- Use bullet points where appropriate
- Use concise paragraphs
- Use properly formatted markdown code blocks
- Keep explanations dense with value but easy to scan

# Important Constraints
- DO NOT SUMMARIZE
- Do not include conversational filler
- Do not include timestamps unless explicitly useful
- Do not mention the existence of the transcript
- Do not repeat the same information multiple times
- Do not oversimplify technical concepts
- Do not hallucinate libraries, APIs, or implementation details
- Preserve technical accuracy over simplification

When code appears in the transcript:
- Preserve all meaningful code examples
- Code is a critical component of the course notes. Provide code where it will enhance understanding.
- Correct obvious syntax issues caused by transcription errors
- Reformat code using clean and professional Python styling
- Add short explanatory comments only when useful
- Explain:
  - what the code does
  - why it matters
  - common pitfalls
  - practical use cases
- Separate conceptual explanations from implementation details

If code is discussed conceptually but not fully shown:
- Reconstruct concise illustrative examples when appropriate
- Do NOT invent large systems or unsupported implementation details
- Stay grounded in the transcript content

# Quality Standard
The final output should resemble polished technical course material. DO NOT PROVIDE A SUMMARY.

The notes should:
- help learners study efficiently
- serve as long-term reference material
- provide implementation clarity
- preserve engineering intuition
- balance theory and practice

# Transcripts
{transcripts}
"""

In [2]:
# prepare module 1 transcripts
from pathlib import Path

In [3]:
# Loop through all filepaths in module 1
module1_path = Path("../1_ingestion/transcripts_module1/")
filepaths = [t for t in module1_path.iterdir()]

In [4]:
from utils import read_json

# Loop through all files, fetch and join all transcripts and assign to one object
transcripts = []

for f in filepaths:
    transcript = read_json(f)[1]
    transcripts.append(transcript)

In [7]:
transcripts_string = "\n".join(transcripts)

In [ ]:
# Try to call the gpt-5-mini

from utils import llm


In [15]:
messages = []

if instructions:
    messages.append({
        "role": "system",
        "content": instructions
    })

messages.append({
    "role": "user",
    "content": transcripts_string
})

print(messages)

[{'role': 'system', 'content': '\n# Role\nYou are a senior technical educator and curriculum designer specializing in Python, AI engineering, and machine learning.\n\n# Objective\nTransform the provided transcript into high-quality educational course notes that are clear, technically accurate, well-structured, and optimized for learning and later review.\n\n# Audience\nThe audience consists of beginner-to-intermediate learners with some technical background who want practical understanding, implementation knowledge, and conceptual clarity.\n\n# Core Responsibilities\nYour task is to:\n- Extract the most important concepts from the transcripts\n- Organize information into coherent sections\n- Remove filler speech, repetition, and conversational noise\n- Preserve all meaningful technical insights\n- Explain concepts clearly and concisely\n- Convert spoken explanations into polished written educational material\n- Preserve important implementation details and engineering reasoning\n- DO N

In [16]:
import json

print(json.dumps(messages, indent=2))

[
  {
    "role": "system",
    "content": "\n# Role\nYou are a senior technical educator and curriculum designer specializing in Python, AI engineering, and machine learning.\n\n# Objective\nTransform the provided transcript into high-quality educational course notes that are clear, technically accurate, well-structured, and optimized for learning and later review.\n\n# Audience\nThe audience consists of beginner-to-intermediate learners with some technical background who want practical understanding, implementation knowledge, and conceptual clarity.\n\n# Core Responsibilities\nYour task is to:\n- Extract the most important concepts from the transcripts\n- Organize information into coherent sections\n- Remove filler speech, repetition, and conversational noise\n- Preserve all meaningful technical insights\n- Explain concepts clearly and concisely\n- Convert spoken explanations into polished written educational material\n- Preserve important implementation details and engineering reaso

In [8]:
llm(user_prompt=transcripts_string, instructions=instructions)

'# 1. Title\n\n## Machine Learning Fundamentals: Supervised Learning, Rule-Based Systems, Model Selection, CRISP-DM, NumPy, and Pandas\n\n---\n\n# 2. Overview\n\nThis session introduces the core ideas and workflow of machine learning through practical examples, especially:\n\n- **Car price prediction** as the first supervised learning example\n- **Spam detection** as a comparison between **rule-based systems** and **machine learning**\n- **Supervised learning** as a formal machine learning setup\n- **Model selection** using training, validation, and test data\n- **CRISP-DM** as a practical framework for organizing ML projects\n- **NumPy** for numerical computation\n- **Linear algebra** operations used in ML\n- **Pandas** for working with tabular data\n\nThe material builds intuition for how machine learning systems are designed, trained, evaluated, and deployed, while also introducing the Python tools needed to implement them.\n\n---\n\n# 3. Key Concepts\n\n## Machine Learning\n- Machi

In [12]:
response_path = Path("results/module1_iteration1.md")
markdown_text = response_path.read_text()
print(markdown_text)

'# 1. Title\n\n## Machine Learning Fundamentals: Supervised Learning, Rule-Based Systems, Model Selection, CRISP-DM, NumPy, and Pandas\n\n---\n\n# 2. Overview\n\nThis session introduces the core ideas and workflow of machine learning through practical examples, especially:\n\n- **Car price prediction** as the first supervised learning example\n- **Spam detection** as a comparison between **rule-based systems** and **machine learning**\n- **Supervised learning** as a formal machine learning setup\n- **Model selection** using training, validation, and test data\n- **CRISP-DM** as a practical framework for organizing ML projects\n- **NumPy** for numerical computation\n- **Linear algebra** operations used in ML\n- **Pandas** for working with tabular data\n\nThe material builds intuition for how machine learning systems are designed, trained, evaluated, and deployed, while also introducing the Python tools needed to implement them.\n\n---\n\n# 3. Key Concepts\n\n## Machine Learning\n- Machi

In [13]:
from IPython.display import Markdown, display

response_text_clean = markdown_text.replace("\\n", "\n")

display(Markdown(markdown_text))

'# 1. Title\n\n## Machine Learning Fundamentals: Supervised Learning, Rule-Based Systems, Model Selection, CRISP-DM, NumPy, and Pandas\n\n---\n\n# 2. Overview\n\nThis session introduces the core ideas and workflow of machine learning through practical examples, especially:\n\n- **Car price prediction** as the first supervised learning example\n- **Spam detection** as a comparison between **rule-based systems** and **machine learning**\n- **Supervised learning** as a formal machine learning setup\n- **Model selection** using training, validation, and test data\n- **CRISP-DM** as a practical framework for organizing ML projects\n- **NumPy** for numerical computation\n- **Linear algebra** operations used in ML\n- **Pandas** for working with tabular data\n\nThe material builds intuition for how machine learning systems are designed, trained, evaluated, and deployed, while also introducing the Python tools needed to implement them.\n\n---\n\n# 3. Key Concepts\n\n## Machine Learning\n- Machine learning learns patterns from data rather than relying on hard-coded rules.\n- A model is the result of training on data and can later make predictions on unseen examples.\n\n## Features and Target\n- **Features (`X`)**: the input information used for prediction.\n- **Target (`y`)**: the output value or label the model learns to predict.\n\n## Supervised Learning\n- A learning setup where training examples include both input features and correct answers.\n- Includes:\n  - **Regression**: predict a number\n  - **Classification**: predict a category\n  - **Ranking**: order items by relevance or preference\n\n## Rule-Based Systems\n- Humans manually define if-then rules.\n- Easy to start with, but hard to maintain as rules grow in number and complexity.\n\n## Model Selection\n- Different models are trained and compared.\n- Validation data is used for choosing the best model.\n- Test data is used once at the end to estimate final performance.\n\n## CRISP-DM\nA six-step process for organizing machine learning projects:\n1. Business understanding\n2. Data understanding\n3. Data preparation\n4. Modeling\n5. Evaluation\n6. Deployment\n\n## NumPy\n- Python library for numerical arrays and matrix operations.\n- Supports vectorized element-wise operations and linear algebra primitives.\n\n## Linear Algebra\n- Vector and matrix operations are foundational for ML.\n- Important operations include:\n  - scalar multiplication\n  - vector addition\n  - dot product\n  - matrix-vector multiplication\n  - matrix-matrix multiplication\n  - identity matrix\n  - matrix inverse\n\n## Pandas\n- Python library for working with tabular data.\n- Main abstraction: **DataFrame**\n- Main column abstraction: **Series**\n\n---\n\n# 4. Detailed Explanations\n\n## 4.1 Machine Learning vs. Rule-Based Systems\n\n### Rule-Based Systems\nIn a rule-based spam filter, humans inspect data manually and encode patterns directly into code.\n\nExamples of rules:\n- If sender is `promotions@online.com`, mark as spam.\n- If subject contains `"tax review"` and sender domain is `online.com`, mark as spam.\n- If body contains `"deposit"`, mark as spam.\n\n### Why this matters\nRule-based systems are intuitive at first, but they become difficult to maintain because:\n- spam patterns change over time\n- each new pattern requires new code\n- rules can conflict with one another\n- fixing one rule may break others\n\n### Limitations\n- brittle\n- manual\n- hard to scale\n- difficult to keep up with changing behavior\n\n---\n\n## 4.2 Machine Learning Approach to Spam Detection\n\nInstead of manually encoding rules, machine learning uses:\n- a collection of labeled emails\n- extracted features from each email\n- a model trained to predict spam vs. not spam\n\n### Typical ML pipeline\n1. Collect data\n2. Extract features\n3. Train model\n4. Use model for prediction\n\n### Why it matters\nML systems can learn patterns automatically from examples and adapt better than hard-coded rules when behavior changes.\n\n---\n\n## 4.3 Features and Labels\n\n### Features\nFeatures are measurable characteristics of the input object.\n\nFor spam detection, examples include:\n- title length\n- body length\n- sender name\n- sender domain\n- presence of words like `"deposit"`\n\nThese can be:\n- **binary features**: true/false, often encoded as `1`/`0`\n- numeric features\n- categorical features\n\n### Target variable\nThe target is the value to predict:\n- `1` for spam\n- `0` for not spam\n\n### Why it matters\nA machine learning model does not learn directly from raw objects; it learns from structured features.\n\n---\n\n## 4.4 Binary Feature Encoding\n\nBinary features take only two values:\n- true / false\n- 1 / 0\n\nExample:\n- `title_longer_than_10 = 1`\n- `body_longer_than_100 = 1`\n- `sender_is_promotions_online = 0`\n- `domain_is_online = 1`\n\n### Engineering consideration\nBinary features are simple and often effective, especially when derived from rule-like intuition.\n\n---\n\n## 4.5 Supervised Learning\n\nSupervised learning means the model is trained with examples that include both:\n- input features `X`\n- correct target values `y`\n\nThe model learns a function:\n\n```text\ng(X) ≈ y\n```\n\nWhere:\n- `X` is the feature matrix\n- `y` is the target vector\n- `g` is the learned model\n\n### Why it matters\nThis is the standard setup for many practical ML tasks where historical labeled data is available.\n\n---\n\n## 4.6 Types of Supervised Learning\n\n### Regression\nPredict a numeric value.\n\nExamples:\n- car price prediction\n- house price prediction\n\nOutput:\n- a number, often continuous\n\n### Classification\nPredict a category.\n\nExamples:\n- spam vs. not spam\n- image of a car vs. not a car\n\n#### Binary classification\nA special case with two classes:\n- spam / not spam\n- fraud / not fraud\n\n#### Multi-class classification\nMore than two classes:\n- car / cat / dog\n- multiple product categories\n\n### Ranking\nPredict an ordering of items by relevance or preference.\n\nExamples:\n- search engine results\n- recommender systems\n- e-commerce product ranking\n\n### Tradeoffs\nDifferent tasks require different output types, metrics, and model behavior.\n\n---\n\n## 4.7 Car Price Prediction Example\n\n### Problem\nA user wants to list a car for sale and needs help choosing a reasonable price.\n\n### Available information\nExamples of useful features:\n- age/year\n- manufacturer\n- model\n- mileage\n- number of doors\n- engine characteristics\n\n### Target\n- price\n\n### How it works\nA model is trained on many examples of cars with known prices. It learns relationships such as:\n- older cars tend to be cheaper\n- high mileage often lowers price\n- some manufacturers are more expensive than others\n\n### Why it matters\nThis shows the standard supervised learning pattern:\n- observed inputs\n- known outputs\n- learned model\n- prediction on new unseen data\n\n---\n\n## 4.8 Spam Detection Example\n\n### Rule-based stage\nRules are created from observed patterns:\n- sender domain\n- subject keywords\n- body keywords\n\n### ML stage\nThose same ideas can become features:\n- sender is a known spam sender\n- subject contains suspicious terms\n- body contains suspicious terms\n\n### Why it matters\nStarting with rules can help identify useful features for machine learning.\n\n---\n\n## 4.9 Model Outputs and Decision Thresholds\n\nFor binary classification, many models output a probability, not a hard class label.\n\nExample outputs:\n- `0.8`\n- `0.6`\n- `0.1`\n- `0.01`\n\nA threshold is used to convert probability into a final decision:\n- if `p >= 0.5`, classify as spam\n- otherwise, classify as not spam\n\n### Practical consideration\nThe threshold does not have to be `0.5`. It can be tuned depending on business goals:\n- lower threshold: catch more spam, but risk more false positives\n- higher threshold: fewer false positives, but more spam reaches inbox\n\n---\n\n## 4.10 Why Model Selection Is Needed\n\nDifferent models may perform differently on the same problem:\n- logistic regression\n- decision trees\n- random forest\n- neural networks\n\nThe goal is to choose the model that performs best on unseen data.\n\n### Important issue: overfitting to validation data\nIf you compare too many models on the same validation set, one model may appear best just by chance.\n\nThis is a **multiple comparison problem**:\n- a model can “get lucky” on a specific validation subset\n- its performance may not generalize\n\n---\n\n## 4.11 Train / Validation / Test Split\n\nA standard workflow is:\n- **training set**: fit the model\n- **validation set**: compare models and tune decisions\n- **test set**: final evaluation only\n\nTypical split example:\n- 60% train\n- 20% validation\n- 20% test\n\n### Why it matters\nThis protects against overestimating model performance due to chance.\n\n### Workflow\n1. Train each candidate model on training data\n2. Evaluate on validation data\n3. Select the best model\n4. Evaluate once on test data\n5. If appropriate, retrain using train + validation and then test once\n\n---\n\n## 4.12 Final Model Training After Selection\n\nAfter selecting the best model, you may combine training and validation data to train the final model.\n\n### Why this matters\nOnce model selection is complete, the validation data no longer needs to be held out for model choice. Using more data can improve the final model.\n\n### Caveat\nThe test set must remain untouched until the final evaluation.\n\n---\n\n## 4.13 CRISP-DM Workflow\n\nCRISP-DM is a structured methodology for organizing ML projects.\n\n### Step 1: Business understanding\n- What problem are we solving?\n- Is machine learning even needed?\n- How do we measure success?\n\n### Step 2: Data understanding\n- What data is available?\n- Is it reliable?\n- Is it large enough?\n- Are labels trustworthy?\n\n### Step 3: Data preparation\n- Clean data\n- Fix missing values\n- Transform raw data into model-ready format\n- Extract features\n\n### Step 4: Modeling\n- Train candidate models\n- Compare them\n- Select the best one\n\n### Step 5: Evaluation\n- Check whether the selected model meets the business goal\n- Confirm results on held-out data\n\n### Step 6: Deployment\n- Put the model into production\n- Monitor performance\n- Maintain reliability and scalability\n\n### Why it matters\nMachine learning projects are not just about training models. They require end-to-end project management.\n\n---\n\n## 4.14 NumPy\n\nNumPy is a Python library for numerical computation.\n\n### Why it matters\nIt provides:\n- arrays\n- fast vectorized operations\n- matrix operations\n- support for linear algebra\n- random number generation\n\n### Common array creation functions\n- `np.zeros`\n- `np.ones`\n- `np.full`\n- `np.array`\n- `np.arange`\n- `np.linspace`\n\n### Vectorized operations\nNumPy applies operations element-wise:\n- add\n- subtract\n- multiply\n- divide\n- power\n- comparisons\n\nThis is much cleaner and faster than writing loops.\n\n### Example\n```python\nimport numpy as np\n\na = np.arange(5)\nb = a * 2 + 10\n```\n\n---\n\n## 4.15 NumPy Indexing and Shapes\n\n### One-dimensional arrays\nAccess elements with standard indexing:\n```python\na[2]\n```\n\n### Two-dimensional arrays\nAccess rows and columns:\n```python\nn[0, 1]\nn[:, 1]\nn[1, :]\n```\n\n### Shape\nThe `shape` attribute gives dimensions.\n\nExamples:\n- 1D array: `(4,)`\n- 2D array: `(5, 2)`\n\n### Why it matters\nShape matching is essential in linear algebra and ML pipelines.\n\n---\n\n## 4.16 Random Number Generation in NumPy\n\nNumPy can generate random arrays:\n- uniform distribution\n- normal distribution\n- random integers\n\n### Reproducibility\nUse a random seed to make results repeatable.\n\n```python\nnp.random.seed(2)\n```\n\n### Why it matters\nReproducibility is crucial in experiments, debugging, and model comparison.\n\n---\n\n## 4.17 Linear Algebra Operations\n\nLinear algebra is the mathematical foundation of many ML algorithms.\n\n### Scalar multiplication\nMultiply every element of a vector by a scalar.\n\n### Vector addition\nAdd corresponding elements of two vectors.\n\n### Dot product\nProduces a single number.\n\nFormula:\n```text\nu · v = Σ u_i v_i\n```\n\n### Why it matters\nDot products are used in:\n- linear regression\n- neural networks\n- similarity measures\n- optimization\n\n---\n\n## 4.18 Matrix-Vector Multiplication\n\nA matrix-vector product computes a dot product between each row of the matrix and the vector.\n\nIf `U` is a matrix with `k` rows and `v` is a vector, then the result is a vector of length `k`.\n\n### Why it matters\nThis operation is one of the most common computations in machine learning.\n\n---\n\n## 4.19 Matrix-Matrix Multiplication\n\nMatrix-matrix multiplication can be understood as multiple matrix-vector multiplications.\n\nIf:\n- `U` is a matrix\n- `V` is a matrix\n\nthen each column of the result is `U` multiplied by one column of `V`.\n\n### Why it matters\nThis decomposition helps explain how matrix multiplication works internally.\n\n---\n\n## 4.20 Identity Matrix\n\nAn identity matrix has:\n- `1`s on the diagonal\n- `0`s elsewhere\n\nMultiplying any compatible matrix by identity leaves it unchanged.\n\n### Why it matters\nThe identity matrix is the matrix analogue of the number `1`.\n\n---\n\n## 4.21 Matrix Inverse\n\nFor a square matrix `A`, the inverse `A^-1` satisfies:\n\n```text\nA * A^-1 = I\n```\n\nwhere `I` is the identity matrix.\n\n### Important limitation\nOnly square matrices can have inverses.\n\n### Why it matters\nMatrix inverses are useful in some ML derivations, especially classical linear models.\n\n---\n\n## 4.22 Pandas\n\nPandas is a Python library for working with tabular data.\n\n### Main abstraction: DataFrame\nA DataFrame is a table with rows and columns.\n\n### Main column abstraction: Series\nEach column in a DataFrame is a Series.\n\n### Why it matters\nMost real-world ML work begins with tabular data cleaning, inspection, transformation, and feature engineering.\n\n---\n\n## 4.23 Creating DataFrames\n\nYou can create DataFrames from:\n- list of lists\n- list of dictionaries\n\nExample:\n\n```python\nimport pandas as pd\n\ndata = [\n    ["BMW", "X5", 2019, 3000, 300, 6, "automatic", "SUV", 65000],\n    ["Toyota", "Corolla", 2018, 1800, 140, 4, "manual", "sedan", 18000],\n]\ncolumns = [\n    "make", "model", "year", "engine_displacement",\n    "horsepower", "num_cylinders", "transmission_type",\n    "vehicle_style", "price"\n]\n\ndf = pd.DataFrame(data, columns=columns)\n```\n\n---\n\n## 4.24 Inspecting DataFrames\n\nCommon first step:\n```python\ndf.head()\n```\n\nThis shows the first rows and is useful for understanding structure and content.\n\n---\n\n## 4.25 Accessing Columns and Rows\n\n### Column access\n```python\ndf.make\ndf["make"]\ndf[["make", "model", "price"]]\n```\n\n### Row access by label\n```python\ndf.loc[1]\ndf.loc[[1, 3]]\n```\n\n### Row access by position\n```python\ndf.iloc[1]\ndf.iloc[[1, 3]]\n```\n\n### Why it matters\nLabel-based and position-based indexing solve different problems.\n\n---\n\n## 4.26 Adding and Removing Columns\n\n### Add column\n```python\ndf["id"] = [1, 2, 3, 4, 5]\n```\n\n### Remove column\n```python\ndel df["id"]\n```\n\n---\n\n## 4.27 Index in Pandas\n\nThe index identifies rows.\n\n### Resetting index\n```python\ndf = df.reset_index(drop=True)\n```\n\n### Why it matters\nWhen filtering or reordering data, keeping row identity consistent is important.\n\n---\n\n## 4.28 Element-wise Operations in Pandas\n\nPandas Series support vectorized arithmetic and comparisons.\n\nExamples:\n```python\ndf["engine_hp"] / 100\ndf["year"] > 2015\ndf[(df["make"] == "Nissan") & (df["year"] > 2015)]\n```\n\n### Why it matters\nThis is the basis for filtering, transformation, and feature engineering.\n\n---\n\n## 4.29 String Operations in Pandas\n\nString operations are available via `.str`.\n\nExamples:\n```python\ndf["vehicle_style"].str.lower()\ndf["vehicle_style"].str.replace(" ", "_")\n```\n\n### Why it matters\nText cleanup is common in data preparation.\n\n---\n\n## 4.30 Summary Statistics in Pandas\n\nUseful methods:\n- `mean()`\n- `max()`\n- `min()`\n- `describe()`\n- `nunique()`\n- `isnull()`\n- `sum()`\n\n### Example\n```python\ndf["price"].describe()\ndf.describe()\ndf["make"].nunique()\ndf.isnull().sum()\n```\n\n### Why it matters\nThese help you inspect distributions, missing values, and basic data quality.\n\n---\n\n## 4.31 Grouping in Pandas\n\nExample: average price by transmission type\n\n```python\ndf.groupby("transmission_type")["price"].mean()\n```\n\n### Why it matters\nGrouping is essential for aggregating data by category, similar to SQL `GROUP BY`.\n\n---\n\n## 4.32 Converting DataFrames to Dictionaries\n\nIf needed, convert to record-oriented dictionaries:\n\n```python\ndf.to_dict(orient="records")\n```\n\n### Why it matters\nUseful for exporting or interfacing with other systems.\n\n---\n\n# 5. Examples\n\n## Example 1: Spam Detection with Rules\n```python\ndef is_spam(email):\n    if email["sender"] == "promotions@online.com":\n        return True\n    if "tax review" in email["subject"].lower() and email["sender_domain"] == "online.com":\n        return True\n    if "deposit" in email["body"].lower():\n        return True\n    return False\n```\n\n### Limitation\nThis becomes fragile as new spam patterns appear.\n\n---\n\n## Example 2: Binary Feature Vector for an Email\n```python\nfeatures = [\n    int(len(subject) > 10),\n    int(len(body) > 100),\n    int(sender == "promotions@online.com"),\n    int(sender_domain == "online.com"),\n    int("deposit" in body.lower()),\n]\n```\n\n---\n\n## Example 3: Train / Validation / Test Split\n```python\n# Pseudocode\ntrain_data = data[:6000]\nval_data = data[6000:8000]\ntest_data = data[8000:]\n```\n\n---\n\n## Example 4: NumPy Array Operations\n```python\nimport numpy as np\n\na = np.arange(5)      # [0, 1, 2, 3, 4]\nb = a * 2             # [0, 2, 4, 6, 8]\nc = a + b             # [0, 3, 6, 9, 12]\nmask = a >= 2         # [False, False, True, True, True]\nfiltered = a[mask]    # [2, 3, 4]\n```\n\n---\n\n## Example 5: Pandas Filtering\n```python\nimport pandas as pd\n\nfiltered = df[(df["make"] == "Nissan") & (df["year"] > 2015)]\n```\n\n---\n\n## Example 6: Grouping by Category\n```python\naverage_price = df.groupby("transmission_type")["price"].mean()\n```\n\n---\n\n# 6. Mermaid Diagram\n\n```mermaid\nflowchart TD\n    A[Raw Data] --> B[Feature Extraction]\n    B --> C[Train Model]\n    C --> D[Validation Set Evaluation]\n    D --> E[Select Best Model]\n    E --> F[Test Set Evaluation]\n    F --> G[Deployment]\n    G --> H[Monitoring]\n    H --> A\n\n    subgraph CRISP-DM\n        I[Business Understanding]\n        J[Data Understanding]\n        K[Data Preparation]\n        L[Modeling]\n        M[Evaluation]\n        N[Deployment]\n    end\n\n    I --> J --> K --> L --> M --> N\n```\n\n---\n\n# 7. Common Pitfalls\n\n- **Hard-coding rules too early**\n  - Works initially but becomes hard to maintain.\n- **Using test data during model selection**\n  - Leads to overly optimistic performance estimates.\n- **Assuming the validation set gives the final answer**\n  - A model may get lucky on validation data.\n- **Ignoring missing values**\n  - Missing data can break model training or skew results.\n- **Not checking feature shapes**\n  - Shape mismatches are a common source of bugs in NumPy and ML code.\n- **Confusing `.loc` and `.iloc` in Pandas**\n  - One is label-based, the other is position-based.\n- **Overwriting data without checking**\n  - Especially when cleaning columns or resetting indexes.\n- **Forgetting reproducibility**\n  - Without a fixed seed, experiments may be hard to reproduce.\n- **Treating the 0.5 threshold as universal**\n  - Threshold should be chosen based on the problem and business cost of errors.\n\n---\n\n# 8. Best Practices\n\n- Start with a simple baseline before building complex systems.\n- Use rule-based analysis to discover useful candidate features.\n- Separate training, validation, and test data strictly.\n- Keep the test set untouched until final evaluation.\n- Use feature engineering to convert raw data into model-ready form.\n- Check data quality early:\n  - missing values\n  - wrong labels\n  - inconsistent text formats\n- Make transformations reproducible and explicit.\n- Use vectorized NumPy/Pandas operations instead of loops when possible.\n- Inspect data immediately after loading with `head()`, `describe()`, and missing-value checks.\n- In ML projects, iterate:\n  - define problem\n  - evaluate data\n  - train model\n  - measure results\n  - improve\n\n---\n\n# 9. Key Takeaways\n\n- Machine learning replaces manually written rules with learned patterns from data.\n- Supervised learning uses labeled examples with features `X` and target `y`.\n- Regression predicts numbers; classification predicts categories; ranking orders items.\n- Spam detection is a good example of the difference between rule-based systems and ML.\n- Model selection requires training, validation, and test splits to avoid overfitting to validation data.\n- CRISP-DM provides a practical end-to-end process for ML projects.\n- NumPy is essential for numerical computation and linear algebra.\n- Pandas is the standard tool for manipulating tabular data in Python.\n- Data preparation and feature engineering are as important as model training.\n- In real ML work, the model is only one part of a larger engineering and product workflow.'

In [14]:
# save clean response
response_path.write_text(markdown_text.replace("\\n", "\n"), encoding="utf-8")


20651

- `Path.write_text()` returns the number of characters written to the file. 
    - 20651 was returned because that number of characters was written to the file.